File: Java/Algorithms/MergeSort.java
package Algorithms;

public class MergeSort {

	private int[] arrClone;
	
	public void sort(int [] arr) {
		arrClone = arr.clone();
		sort(arr, 0, arr.length - 1);
	}
	
	private void sort(int[] arr, int low, int high) {
		
		if(low < high) {
			int mid = (low + high) / 2;
			
			sort(arr, low, mid);
			sort(arr, mid + 1, high);
			
			merge(arr, low, mid, high);	
		}
	}
	
	private void merge(int[] arr, int low, int mid, int high) {
		int i,j,k;
		i = low;
		j = mid + 1;
		
		for(k = low; k <= high; k++) {
			arrClone[k] = arr[k];
		}
		
		for(k = low; k <= high; k++) {
			if ( i > mid ) {
				arr[k] = arrClone[j];
				j++;
			}
			else if ( j > high ) {
				arr[k] = arrClone[i];
				i++;
			}
			else if ( arrClone[i] > arrClone[j] ) {
				arr[k] = arrClone[j];
				j++;
			}
			else {
				arr[k] = arrClone[i];
				i++;
			}
		}
	}
	
	private static void print(int[] arr) {
		String res = "[";
		for(int i = 0; i< arr.length; i++) {
			res += arr[i];
			if(i < arr.length - 1) {
				res += ", ";
			}
		}
		res +="]";
		System.out.println(res);
	}
	
	public static void main(String[] args) {
		// Array to sort.
		MergeSort m = new MergeSort();
		int[] arr = {10,9,8,7,5,6,4,3,2,1};
		
		m.sort(arr);
		print(arr);
		
	}
	
	
}

File: TypeScript/Algorithms/MergeSort.ts
class MergeSort
{
    arrClone : number[];

    public sort(arr: number[])
    {
        // Clone the array for sorting.
        this.arrClone = Object.assign([], arr);
        this.sortAux(arr, 0, arr.length - 1);

    }

    private sortAux(arr: number[], low: number, high: number)
    {
        if(low < high)
        {
            const mid = Math.floor((low + high) / 2);

            this.sortAux(arr, low, mid);
            this.sortAux(arr, mid + 1, high);

            this.merge(arr, low, mid, high);

        }
    }

    private merge(arr: number[], low: number, mid: number, high: number)
    {
        let i = low;
        let j = mid + 1;
        let k = 0;

        for(k = low; k <= high; k++)
        {
            this.arrClone[k] = arr[k];
        }

        for(k = low; k <= high; k++)
        {
            if( i > mid)
            {
                arr[k] = this.arrClone[j];
                j++;
            }
            else if(j > high)
            {
                arr[k] = this.arrClone[i];
                i++;
            }
            else if(this.arrClone[i] > this.arrClone[j])
            {
                arr[k] = this.arrClone[j];
                j++;
            }
            else
            {
                arr[k] = this.arrClone[i];
                i++;
            }
        }
    }
}